src/combinatorics/simple_graph/inc_matrix.lean

```diff
@@ -0,0 +1,360 @@
+and provides theorems and lemmas connecting graph properties to computational properties of the
+matrix. It also defines the notion of `orientation` for a `simple_graph`, picking a direction for
+each undirected edge in the graph and then defining the oriented incidence matrix
+`oriented_inc_matrix` based on that.
+
+## Main definitions
+
+* `inc_matrix`: The incidence matrix `M` of a `simple_graph` with coefficients in a given ring R.
+* `orientation`: A structure that defines a choice of direction on the edges of a `simple_graph`.
+* `oriented_inc_matrix`: The oriented incidence matrix `N(o)` of a `simple_graph` with
+respect to a given `orientation`.
+
+## Main results
+
+* `inc_matrix_element_power_id`: Every element from `M` is idempotent.
+* `degree_equals_sum_of_incidence_row`: The sum of elements from row `i` of `M` is equal to the
+`degree` of vertex `i`.
+* `oriented_inc_matrix_elem_squared`: The square of each element from `N(o)` is equal to the
+corresponding element from `M`.
+* `vec_mul_oriented_inc_matrix`: `(xᵀ ⬝ N(o)) e = x o.head(e) - x o.tail(e)`.
+
+## References
+
+<https://en.wikipedia.org/wiki/Orientation_(graph_theory)>
+-/
+
+open_locale big_operators matrix
+open finset matrix simple_graph sym2
+
+universes u v
+variables {R : Type u} {V : Type v}
+
+section to_move
+
+@[simp]
+lemma ite_mul_ite_zero_right
+  {R : Type*} [mul_zero_class R] {P Q : Prop} [decidable P] [decidable Q] (a b : R):
+  (ite P a 0) * (ite Q b 0) = ite (P ∧ Q) (a * b) 0 :=
+by simp [←ite_and]
+
+end to_move
+
+namespace simple_graph
+
+variables (G : simple_graph V)
+
+/-! ### Helping lemmas for edges -/
+
+lemma incidence_equiv {i : V} {e : G.edge_set} : ↑e ∈ G.incidence_set i ↔ i ∈ (↑e : sym2 V) :=
+by simp only [incidence_set, true_and, set.mem_sep_eq, subtype.coe_prop]
+
+lemma mem_incidence_set_iff {u v w : V} :
+  ⟦(u, v)⟧ ∈ G.incidence_set w ↔ ⟦(u, v)⟧ ∈ G.edge_set ∧ (w = u ∨ w = v) :=
+by { convert iff.rfl, simp only [mem_iff] }
+
+lemma edge_in_two_incidence_sets {i j : V} {e : sym2 V} (H_ne : i ≠ j) :
+  e ∈ G.incidence_set i ∧ e ∈ G.incidence_set j → e = ⟦(i, j)⟧ :=
+begin
+  refine quotient.rec_on_subsingleton e (λ p, _),
+  rcases p with ⟨v, w⟩,
+  rintro ⟨⟨-, H_i⟩, ⟨-, H_j⟩⟩,
+  rw ←elems_iff_eq H_ne,
+  tauto
+end
+
+lemma mem_incidence_sets_iff_eq {i j : V} {e : sym2 V} (h : G.adj i j) :
+  e ∈ G.incidence_set i ∧ e ∈ G.incidence_set j ↔ e = ⟦(i, j)⟧ :=
+begin
+ split,
+  { apply edge_in_two_incidence_sets _,
+  rintro rfl,
+  exact G.loopless i h },
+  { rintro rfl,
+  nth_rewrite 1 eq_swap,
+  simp only [h, G.sym h, mem_incidence_set, and_self] }
+end
+
+lemma adj_iff_exists_edge_coe {i j : V} : G.adj i j ↔ ∃ (e : G.edge_set), ↑e = ⟦(i, j)⟧ :=
+by simp only [mem_edge_set, exists_prop, set_coe.exists, exists_eq_right, subtype.coe_mk]
+
+/-! ## Incidence matrix M -/
+
+section incidence
+
+variables [fintype V] [decidable_eq V] [decidable_rel G.adj] (R) [ring R]
```

Comment:
```suggestion
variables [fintype V] [decidable_eq V] [decidable_rel G.adj] (R) [has_zero R] [has_one R]
```
and for the lemmas that require addition, I think `[semiring R]` should be enough.
Also, you shouldn't need all those `[char_zero R]` throughout.